### PR TITLE
(templating) add sort order option

### PR DESCRIPTION
**Pull request by mtanda**
- https://github.com/grafana/grafana/issues/2839

### CODE REVIEW

**Comment by torkelo** [comment]
this looks great, just needs unit tests

**Comment by mtanda** [comment]
Thanks for review! Add test!!

**Comment by lark** [comment]
Does this PR solve the problem as the screenshot below?

image

The graph is generated using this setting

image

The very strange part is "01 03 02", in every graph. We deliberately using leading 0 padding to make sorting to work as expected both numerical or alphabetical, but we failed badly.

**Comment by mtanda** [comment]
@lark This PR doesn't solve your problem.
This PR add sort order option for template variables.
Your problems is sort order of legend.

So, please create Issue for your problem. (It might be Feature Request.)

**Comment by lark** [comment]
@mtanda Thanks

**Comment by torkelo** [comment]
maybe this needs to also handle options to control descending or ascending (probably in same dropdown option )

**Comment by mtanda** [comment]
@torkelo add descending sort support.

**Comment by mtanda** [comment]
rebase to latest master, then test passed.

**Comment by bergquist** [comment]
Looks good! But I think this code https://github.com/grafana/grafana/pull/5129/files#diff-06b5e9ee4bae9b6297d7565332a15619R362 is a little bit to complex. Would it be possible to make it more simple?

**Comment by mtanda** [comment]
@bergquist little simplify the code. "Math.ceil()" looks like little oddness, but it is required to reuse "reverse sort" code.

**Comment by rajatjindal83** [comment]
@mtanda will this work for templating feature too? Also, can you please add sort order option too (asc vs desc).

With those two, I think it is better PR then the one I submitted (which was more focused on fixing my issue on hand).

Thanks
Rajat Jindal

**Comment by mtanda** [comment]
@rajatjindal83 yes this works for templating feature.
And there are "asc" order. https://github.com/grafana/grafana/pull/5129/files#diff-7f607724d037f020f0255695f2ba9044R40

I think "Without sort" is fit for your use case.
Sort result by Elasticsearch, and then use the result without sort in Grafana.

File: public/app/features/templating/templateValuesSrv.js [hunk 1]
```diff
@@ -355,10 +355,33 @@ function (angular, _, kbn) {
...
       }
+      options = _.uniq(options, 'value');
 
-      return _.sortBy(options, 'text');
+      if (variable.sort === 0) {
+        return options;
+      }
+
+      var sortType = Math.ceil(variable.sort / 2);
+      var reverseSort = (variable.sort % 2 === 0);
```
Comment (rajatjindal83) [inline public/app/features/templating/templateValuesSrv.js]: is it suppose to be variable.sortOptions instead of variable.sort?

Comment (rajatjindal83) [inline public/app/features/templating/templateValuesSrv.js]: nevermind, read more carefully.

**Comment by mtanda** [comment]
@bergquist Is this merged to 4.0 release?

**Comment by bergquist** [comment]
@mtanda Ill look into it.

**Comment by mtanda** [comment]
Thanks!

**Comment by RichiH** [comment]
Thanks @mtanda & @bergquist :)

**Comment by mtanda** [comment]
Thanks for merging!
The Grafana 4.0 becomes more attractive for me!!

**Comment by andyxning** [comment]
@mtanda Does this PR can solve this situation:
* Grafana + InfluxDB
* template variable defined by
  * SHOW TAG VALUES FROM "uptime" WITH KEY = "container_name" WHERE labels =~ /^(name|.*;name):(default;.*|default)$/
  * The result it ordered by time by default and i want this behavior.

**Comment by mtanda** [comment]
@andyxning This should solve your problem.
You can test this feature at demo site.
http://play.grafana.org/?editview=templating

**Comment by andyxning** [comment]
@mtanda Thanks for you reply and i will  have a try. 
Can you please add a new option for supporting this situation:
* As InfluxDB can not work with `SHOW TAG VALUES` and `ORDER BY time`, so we can not make the result in descending order. It seems by default it is in ascending order. I am wondering if you can add a new option "reverse" for reversing the order of the result according to neither alphabetical nor numerical, just reverse the order.

**Comment by mtanda** [comment]
hmm, it should be easy but I'm not sure the suggest will be accepted.
Adding some line https://github.com/grafana/grafana/pull/5129/files#diff-06b5e9ee4bae9b6297d7565332a15619R365, you can add the feature.
Please try it yourself :-)

**Comment by Demfier** [comment]
Hello All,
Can someone please provide docs on how to use this feature?
I am new to using grafana and this feature would be very useful for my usecase.

**Comment by daniellee** [comment]
@Demfier do you mean the sort options for template variables? 

It is found in the query options section when editing a template variable (see below). The docs for it are [here](http://docs.grafana.org/reference/templating/#query-options).

image

**Comment by Demfier** [comment]
@daniellee Thanks for the reply, but this is not what I'm exactly looking for.
I have `table name` as a metric that contains `table size` after one level of drill down. What I am looking for is to be able to sort the table names appearing on basis of `table size`. Any clue how to do that?

Really appreciate your response.

**Comment by daniellee** [comment]
@Demfier That is not possible unfortunately. If you have more questions rather than reviving old issues or PR's, you can ask questions on the [community site](https://community.grafana.com/).

**Comment by justinTM** [comment]
is there any way to sort by order of most recent in time?

**Pull request merged**